File: src/main/java/com/paragon/client/systems/module/impl/misc/Notifier.java
package com.paragon.client.systems.module.impl.misc;

import com.paragon.Paragon;
import com.paragon.api.event.client.ModuleToggleEvent;
import com.paragon.api.event.combat.PlayerDeathEvent;
import com.paragon.api.event.combat.TotemPopEvent;
import com.paragon.client.managers.CommandManager;
import com.paragon.client.managers.notifications.Notification;
import com.paragon.client.managers.notifications.NotificationManager;
import com.paragon.client.managers.notifications.NotificationType;
import com.paragon.client.systems.module.Module;
import com.paragon.client.systems.module.Category;
import com.paragon.client.systems.module.setting.Setting;
import me.wolfsurge.cerauno.listener.Listener;
import net.minecraft.util.text.TextFormatting;

/**
 * @author Wolfsurge
 */
public class Notifier extends Module {

    public static Notifier INSTANCE;

    public static Setting<Boolean> moduleEnabled = new Setting<>("Module Toggle", false)
            .setDescription("Notifies you when you toggle a module");

    public static Setting<Boolean> pop = new Setting<>("Pop", true)
            .setDescription("Notifies you when a player pops a totem");

    public static Setting<Boolean> death = new Setting<>("Death", true)
            .setDescription("Notifies you when a player dies");

    public static Setting<Boolean> noPops = new Setting<>("No Pops", true)
            .setDescription("Notifies you even if the player hasn't popped any totems")
            .setParentSetting(death);

    public Notifier() {
        super("Notifier", Category.MISC, "Notifies you when events happen");

        INSTANCE = this;
    }

    @Listener
    public void onModuleToggle(ModuleToggleEvent moduleToggleEvent) {
        if (moduleEnabled.getValue()) {
            if (!moduleToggleEvent.getModule().isIgnored()) {
                Paragon.INSTANCE.getNotificationManager().addNotification(new Notification("Toggled " + moduleToggleEvent.getModule().getName(), moduleToggleEvent.getModule().getName() + " was " + (moduleToggleEvent.getModule().isEnabled() ? "Enabled" : "Disabled"), NotificationType.INFO));
            }
        }
    }

    @Listener
    public void onTotemPop(TotemPopEvent event) {
        if (pop.getValue()) {
            Paragon.INSTANCE.getNotificationManager().addNotification(new Notification("Totem Pop", event.getPlayer().getName() + " has popped " + Paragon.INSTANCE.getPopManager().getPops(event.getPlayer()) + " totems!", NotificationType.INFO));
        }
    }

    @Listener
    public void onPlayerDeath(PlayerDeathEvent event) {
        if (death.getValue()) {
            if (!noPops.getValue() && event.getPops() == 0) {
                return;
            }

            Paragon.INSTANCE.getNotificationManager().addNotification(new Notification("Player Died", event.getEntityPlayer().getName() + " has died after popping " + event.getPops() + " totems!", NotificationType.INFO));
        }
    }

}

File: src/main/kotlin/com/paragon/client/systems/module/impl/misc/Notifier.kt
package com.paragon.client.systems.module.impl.misc

import com.paragon.Paragon
import com.paragon.api.event.client.ModuleToggleEvent
import com.paragon.api.event.combat.PlayerDeathEvent
import com.paragon.api.event.combat.TotemPopEvent
import com.paragon.api.module.Category
import com.paragon.api.module.Module
import com.paragon.api.setting.Setting
import com.paragon.bus.listener.Listener
import com.paragon.client.managers.notifications.Notification
import com.paragon.client.managers.notifications.NotificationType
import com.paragon.client.systems.module.hud.HUDModule

/**
 * @author Surge
 */
object Notifier : Module("Notifier", Category.MISC, "Notifies you when events happen") {

    private val moduleEnabled = Setting(
        "ModuleToggle",
        false
    ) describedBy "Notifies you when you toggle a module"

    private val pop = Setting(
        "Pop",
        true
    ) describedBy "Notifies you when a player pops a totem"

    private val death = Setting(
        "Death",
        true
    ) describedBy "Notifies you when a player dies"

    private val noPops = Setting(
        "NoPops",
        true
    ) describedBy "Notifies you even if the player hasn't popped any totems" subOf death

    @Listener
    fun onModuleToggle(moduleToggleEvent: ModuleToggleEvent) {
        if (!moduleEnabled.value || !(!moduleToggleEvent.module.isIgnored && moduleToggleEvent.module !is HUDModule)) {
            return
        }

        Paragon.INSTANCE.notificationManager.addNotification(
            Notification(
                moduleToggleEvent.module.name + " was " + if (moduleToggleEvent.module.isEnabled) "Enabled" else "Disabled",
                NotificationType.INFO
            )
        )
    }

    @Listener
    fun onTotemPop(event: TotemPopEvent) {
        if (!pop.value) {
            return
        }

        Paragon.INSTANCE.notificationManager.addNotification(
            Notification(
                event.player.name + " has popped " + Paragon.INSTANCE.popManager.getPops(
                    event.player
                ) + " totems!", NotificationType.INFO
            )
        )
    }

    @Listener
    fun onPlayerDeath(event: PlayerDeathEvent) {
        if (!death.value || (!noPops.value && event.pops == 0)) {
            return
        }
        Paragon.INSTANCE.notificationManager.addNotification(
            Notification(
                event.entityPlayer.name + " has died after popping " + event.pops + " totems!",
                NotificationType.INFO
            )
        )
    }

}